File: Leetcode/Algorithm/java/00135-Candy.java
public class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int candy[] = new int[n];
        candy[0] = 1;
        for (int i = 1; i < n; i++) {
            if (ratings[i - 1] < ratings[i]) {
                candy[i] = candy[i - 1] + 1; 
            } else {
                candy[i] = 1;
            }
        }
        
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i + 1] < ratings[i]) {
                candy[i] = Math.max(candy[i + 1] + 1, candy[i]);
            }
        }
        
        int ans = 0;
        for (int i: candy) {
            ans += i;
        }
        return ans;
    }
}

File: Leetcode/Algorithm/python/00135-Candy.py
class Solution:
    # @param ratings, a list of integer
    # @return an integer
    def candy(self, ratings):
        n = len(ratings)
        candies = [1 for i in xrange(n)]
        
        for i in xrange(1, n):
            if ratings[i] > ratings[i - 1]:
                candies[i] = candies[i - 1] + 1
        for i in reversed(xrange(0, n - 1)):
            if ratings[i] > ratings[i + 1]:
                candies[i] = max(candies[i], candies[i + 1] + 1)
                
        return sum(candies)
